README

XInvaders 3D v1.4.0
-=-=-=-=-=-=-=-=-=-
XInvaders 3D is a vector-graphics Space Invaders clone
for the X Window System.

================================================================

Objective:
-=-=-=-=-=
Clear all the aliens in each level. The UFO yields bonus
points if hit. A new life will be added every 7500 points.

Game-over: either if the player has lost all its lives
or if the alien formation reaches the player position.

================================================================

Point System:
-=-=-=-=-=-=-

From bottom to top:
Row 0 aliens yield 10  pts
Row 1 aliens yield 50  pts
Row 2 aliens yield 100 pts
Row 3 aliens yield 150 pts
Row 4 aliens yield 200 pts

UFO yields 500 pts

================================================================

Keys:
-=-=-

SPACE         : Fire Missile
Left Arrow    : move left
Right Arrow   : move right
Up Arrow      : move up
Down Arrow    : move down
q             : Reset game
f             : Show frame-rate
p             : Pause game
ESC           : Terminate Program

================================================================

How to compile and play the game:
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

1. Requirements: gcc and Xlib
2. If you have xinv3d-dist-{version}.tar.gz, skip to step 5.
3. run autoreconf -i
4. run automake
5. run ./configure (options)
6. Compile the game, type: make
7. Run game, type: xinv3d or select menu: Games/Arcade/xinv3d

note:
For Makefile.dos or Makefile.win you may need to edit the
Makefile.xxx depending on where your headers and libraries are.
For those systems please see the README.xxx for that system.

================================================================

Original Author and Original Website:
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
Don Llopis
del23@users.sourceforge.net
http://xinvaders3d.sourceforge.net

================================================================

Original Greetings (from Don Llopis):
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
Greets to following people whose games helped inspire
this project:

Bill Kendrick
who wrote ICBM3d http://www.newbreedsoftware.com

Yuusuke Hashimoto
who wrote XSoldier http://www.surfline.ne.jp/hachi/xsoldier.html

James Eibisch
who wrote a Space Invaders clone in quick-basic
http://www.revolver.demon.co.uk/coding/index.html

Thanks to the following people who contributed to the development
of XInvaders3d:

Thomas Boutell for contributing the Windows Port of XInvaders, and
misc code fixes.

Sam Bushell for the ZONE_HEIGHT bug fix.

Andreas Jeitler for creating the RPM package.

Chris Butler for creating the DEB package.

Peter da Silva for the FreeBSD fix.

Bill Kendrick for the Window Manager fix.

================================================================

Current Maintainer and Current Website:
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
Joe Da Silva
https://github.com/JoesCat/xinv3d/releases

================================================================

History:
-=-=-=-=-

1.4.0
      Updated xinv3d to get built using autoreconf -i and automake.
      Added freedesktop and new menu icons (see debian issue #738066).
      Removed compat file (see debian issue #965904).
      Added patch by Helmut Grohne to fix FTCBFS (unreleased 1.3.6-6.1).

1.3.6:
      Minor fixes.
      Changed some game text which had the wrong
      contact information.
      Made some more code clean ups.

1.3.5:
      Moved all game text to gtext.h.
      Removed `goto' in init code.
      Put a space in ``game over.''

1.32-1.34:
      Minor fixes.

1.31:
      Fixed drawing order. Objects were being drawn in reverse.

1.3:
      Many thanks to Thomas Boutell, he has contributed a
      Windows port of XInvaders 3D. See README.win for more
      information.

      XInvaders 3D DOS port finished. XInvaders 3D will now run
      as a full-screen 640x480x8 DOS game. For windows users it
      will run in a DOS window just fine. I used DJGPP+Allegro
      to port it. See README.dos for more information.

      Added Vector-fonts. All text graphics have been
      replaced with Vector-fonts. Gives the game a
      cleaner look.

      New player blinking routine. Player now cycles
      through red colors, rather than being drawn
      each alternate frame, when a new level
      has been started.

      New special effect: 1-UP indicator. When 7500pts
      have been reached a new life is indicated by
      a nice little vector 1-UP floating towards
      the player.

      New alien-shoot missile routine. Aliens now shoot
      more.

      New Distance-of-Point-to-Line routine, used for
      collision detection.

      Added fix for FreeBSD systems, thanks to Peter da Silva.

      Fixed ZONE_HEIGHT bug in game.c, which was generating
      bad pointers, thanks to Sam Bushell.

      Cleaned up code, for portability reasons, thanks to
      Thomas Boutell for his help.

      Left-Shift key is no longer being used.

1.2:
      Cleaned up some more code.

      Adjusted formation movement.

      Changed missile color to yellow, suggested by Mattia Engdeg.

      Fixed some warnings which the pedantic flag generates,
      thanks to Jarmo Pussinen for pointing them out.

1.1:
      Fixed incompatible pointer warnings.

      Fixed window manager bug which Bill Kendrick pointed out.

      Windows were opening at an odd width and height under
      the Enlightenment Window manager.

1.0:
      Initial release.


Changelog
xinv3d (1.5.0)

  * Timing bug for very fast computers fixed. Original speeds restored.
  * Fixed empty stars field in lower-right quadrant before game starts.

 -- Joe Da Silva <digital@joescat.com>  Tue, 22 Feb 2022 08:36:00 -0800

xinv3d (1.4.0)

  * Debian Orphaned xinv3d brought up to date and is buildable with current PCs.
  * Updated xinv3d to get built using autoreconf -i and automake.
  * removed compat file (mentioned in debian issue #965904).
  * Added freedesktop and new icons (mentioned in debian issue #738066).

 -- Joe Da Silva <digital@joescat.com>  Mon, 27 Sep 2021 09:36:00 -0800

xinv3d (1.3.6-6.1) UNRELEASED; urgency=medium

  * Non-maintainer upload.
  * Fix FTCBFS: (Closes: #-1)
    + Let dh_auto_build pass cross tools to make.
    + Defer stripping to dh_strip.

 -- Helmut Grohne <helmut@subdivi.de>  Tue, 27 Nov 2018 19:34:01 +0100

xinv3d (1.3.6-6) unstable; urgency=low

  * Updated build dependency on x-dev to x11proto-core-dev (closes:
    #515418)
  * Added DEB_BUILD_OPTIONS tests to debian/rules
  * Bumped Standards-Version to 3.8.0
  * Added copyright statement to debian/copyright (fix lintian warning)

 -- Chris Butler <chrisb@debian.org>  Sun, 15 Feb 2009 20:06:56 +0000

xinv3d (1.3.6-5) unstable; urgency=low

  * Updated menu entry for new menu policy
  * Updated my address in debian/copyright
  * Bumped Standards-Version to 3.7.2 (no changes required)
  * Updated debhelper usage to latest standards
  * Fixed Makefile so clean doesn't fail when already clean

 -- Chris Butler <chrisb@debian.org>  Thu,  5 Jul 2007 13:25:59 +0100

xinv3d (1.3.6-4) unstable; urgency=low

  * Bumped Standards-Version to 3.6.2
  * Updated build-depends to get rid of obsolete xlibs-dev package
  * Quoted strings in debian/menu to get rid of lintian warning

 -- Chris Butler <chrisb@debian.org>  Mon, 16 Jun 2003 21:45:37 +0100

xinv3d (1.3.6-3) unstable; urgency=low

  * Incorporated changes from NMU, and bumped Standards-Version to 3.5.9
    (closes: #170187)

 -- Chris Butler <chrisb@debian.org>  Sun,  6 Apr 2003 19:21:42 +0100

xinv3d (1.3.6-2.1) unstable; urgency=low

  * Non-Maintainer Upload
  * debian/control:
    - Bumped Standards-Version to 3.5.8
    - Replaced xlib6g-dev build dependency with xlibs-dev. Closes: #170187
    - Updated debhelper build-dependency to version 4
  * debian/rules: switched DH_COMPAT to 4
  * debian/copyright: removed dh_make boilerplate in authors
    description.

 -- Jerome Marant <jerome@debian.org>  Sat,  8 Mar 2003 13:39:29 +0100

xinv3d (1.3.6-2) unstable; urgency=low

  * Remove CVS directories from source package
  * Add missing build dependancy on debhelper (closes: #77880)

 -- Chris Butler <chrisb@debian.org>  Thu, 15 Nov 2001 00:29:39 +0000

xinv3d (1.3.6-1) unstable; urgency=low

  * New upstream release

 -- Chris Butler <chrisb@debian.org>  Wed, 22 Nov 2000 18:29:24 +0000

xinv3d (1.3.1-1) unstable; urgency=low

  * New upstream release
  * debian/menu: fix typo - "/usr/bin/xinv3d" -> "/usr/games/xinv3d"
    (closes: #58844)

 -- Chris Butler <chrisb@sandy.force9.co.uk>  Mon, 13 Mar 2000 20:16:22 +0000

xinv3d (1.3-1) unstable; urgency=low

  * New upstream release

 -- Chris Butler <chrisb@sandy.force9.co.uk>  Sun,  6 Feb 2000 14:48:33 +0000

xinv3d (1.2-1) unstable; urgency=low

  * Initial Release.

 -- Chris Butler <chrisb@sandy.force9.co.uk>  Sun,  6 Feb 2000 14:24:06 +0000


effects.c
/*------------------------------------------------------------------
  effects.c:

    XINVADERS 3D - 3d Shoot'em up
    Copyright (C) 2000 Don Llopis

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

------------------------------------------------------------------*/
#include "game.h"

/*------------------------------------------------------------------
 * Stars
 *
 *
------------------------------------------------------------------*/

enum stars_enum
{
   MAX_STARS = 100
};

struct STARSSTRUCT
{
   VECTOR4      pos;
   VECTOR4      thrustz;
   int          thrust;
   unsigned int color;
} stars[MAX_STARS];

VECTOR4 star_start     = { 0.0f, 0.0f, -2000.0f, 1.0f };
VECTOR4 star_thrust[4] = {{ 0.0f, 0.0f, 50.0f, 1.0f },
                         { 0.0f, 0.0f, 25.0f, 1.0f },
                         { 0.0f, 0.0f, 15.0f, 1.0f },
                         { 0.0f, 0.0f, 10.0f, 1.0f } };


void Stars_init ( void )
{
   int i, m;
   for ( i=0; i<MAX_STARS; i++ )
   {
      Vector_copy ( star_start, stars[i].pos );

      stars[i].pos[XPOS] = (float)((rand () % 1600) - 800);
      stars[i].pos[YPOS] = (float)((rand () % 1600) - 800);

      m = rand () % 4;
      stars[i].thrust = m;
      stars[i].color = WHITE - (m*15);
   }
}

void Stars_draw ( MATRIX4 r )
{
   VECTOR4 tmp;
   int i, j, m, p[2];

   for ( i=0; i<MAX_STARS; i++ )
   {
      j = stars[i].thrust;
      stars[i].pos[ZPOS] += star_thrust[j][ZPOS] * gv->fadjust;

      if ( stars[i].pos[ZPOS] > -1.0f )
      {
         m = rand() % 4;
         stars[i].thrust = m;
         stars[i].color = WHITE - (m*15);
         stars[i].pos[ZPOS] = -2000.0f;
      }
      Matrix_vec_mult ( r, stars[i].pos, tmp );
      Camera_project_point ( tmp, p );
      Draw_point ( p[0], p[1], stars[i].color );
   }
}

/*------------------------------------------------------------------
 * Explosions
 *
 *
------------------------------------------------------------------*/

#define EXPLOSION_ROT 0.23f

enum explosions_enum
{
   MAX_EXPLOSIONS        = 10,
   MAX_PARTICLES         = 4,
   EXPLOSIONS_LIFE       = 1500, /* 1.5 sec */
   EXPLOSION_BLEND_TIME  = 375,  /* 0.375 sec */
   EXPLOSION_COLOR_INC   = 15
};

struct EXPLOSIONSTRUCT
{
   VECTOR4  pos[MAX_PARTICLES];
   int  thrust[MAX_PARTICLES];
   long frame;
   long blend;
   int  color;
   int  active;
} explosions[MAX_EXPLOSIONS];

static int   ecur;   /* EXPLOSION index */
static int   ecount; /* EXPLOSION count */
static int   pcur;   /* THURST index */
static float erot;   /* EXPLOSIONS rotation */

/* shard/particle thrust vectors */
VECTOR4 pthrust[8] = { {-2.0f, 0.0f, 0.0f, 1.0f},
                      {0.0f, -2.0f, 0.0f, 1.0f},
                      {2.0f, 0.0f, 0.0f, 1.0f},
                      {0.0f, 2.0f, 0.0f, 1.0f},
                      {-2.0f, 2.0f, 0.0f, 1.0f},
                      {2.0f, -2.0f, 0.0f, 1.0f},
                      {-2.0f, -2.0f, 0.0f, 1.0f},
                      {2.0f, 2.0f, 0.0f, 1.0f} };


VECTOR4 shard[3] = { {-5.0f, 0.0f, 0.0f, 1.0f},
                    { 0.0f, 5.0f, 0.0f, 1.0f},
                    { 5.0f, -5.0f, 0.0f, 1.0f}};


void Explosions_clear ( void )
{
   int i, j;

   ecur = 0;
   ecount = 0;
   pcur = 0;

   erot = 0.0f;

   for ( i=0; i<MAX_EXPLOSIONS; i++ )
   {
      explosions[i].active = FALSE;
      explosions[i].frame = 0L;
      for (j=0; j<MAX_PARTICLES; j++ )
         Vector_init ( explosions[i].pos[j] );
   }
}

int Explosions_count ( void )
{
   return ecount;
}

void Explosions_add ( OBJECT *obj )
{
   explosions[ecur].active = TRUE;
   explosions[ecur].frame = 0;
   explosions[ecur].color = GREEN;
   explosions[ecur].blend = 0;

   /* ok there are currently only 4 */
   Vector_copy ( obj->pos, explosions[ecur].pos[0] );
   Vector_copy ( obj->pos, explosions[ecur].pos[1] );
   Vector_copy ( obj->pos, explosions[ecur].pos[2] );
   Vector_copy ( obj->pos, explosions[ecur].pos[3] );

   explosions[ecur].thrust[0] = pcur;
   explosions[ecur].thrust[1] = pcur+1;
   explosions[ecur].thrust[2] = pcur+2;
   explosions[ecur].thrust[3] = pcur+3;

   ecur++;
   pcur += 4;
   ecount++;

   if ( ecur > MAX_EXPLOSIONS-1 )
      ecur = 0;

   if ( pcur == 8 )
      pcur = 0;
}

void Explosions_draw ( MATRIX4 r )
{
   int i, j, k, p0[6];
   VECTOR4 tmp[3], shard_tmp[3];
   MATRIX4 tmp_mat, tmp_mat2, erot_mat;

   erot += EXPLOSION_ROT * gv->fadjust;
   Matrix_x_rot ( tmp_mat, erot );
   Matrix_z_rot ( tmp_mat2, erot );
   Matrix_mult  ( tmp_mat, tmp_mat2, erot_mat );
   Matrix_vec_multn ( erot_mat, shard, shard_tmp, 3 );

   for ( i=0; i<MAX_EXPLOSIONS; i++ )
   {
      if ( explosions[i].active )
      {
         explosions[i].frame += gv->msec;
         if ( explosions[i].frame > EXPLOSIONS_LIFE )
         {
            explosions[i].active = FALSE;
            ecount--;
         }
         explosions[i].blend += gv->msec;
         if ( explosions[i].blend > EXPLOSION_BLEND_TIME )
         {
            explosions[i].blend -= EXPLOSION_BLEND_TIME;
            explosions[i].color -= EXPLOSION_COLOR_INC;
         }

         for ( j=0; j<MAX_PARTICLES; j++ )
         {
            k = explosions[i].thrust[j];
            explosions[i].pos[j][XPOS] += pthrust[k][XPOS] * gv->fadjust;
            explosions[i].pos[j][YPOS] += pthrust[k][YPOS] * gv->fadjust;
            explosions[i].pos[j][ZPOS] += pthrust[k][ZPOS] * gv->fadjust;

            Matrix_vec_mult ( r, explosions[i].pos[j], tmp[0] );
            Matrix_copy ( r, tmp_mat );
            Matrix_set_trans ( tmp_mat, tmp[0] );

            Matrix_vec_multn ( tmp_mat, shard_tmp, tmp, 3 );
            Camera_project_points ( tmp, p0, 3 );

            Draw_line ( p0[0], p0[1], p0[2], p0[3], explosions[i].color );
            Draw_line ( p0[2], p0[3], p0[4], p0[5], explosions[i].color );
            Draw_line ( p0[4], p0[5], p0[0], p0[1], explosions[i].color );
         }
      }
   }
}

/*------------------------------------------------------------------
 * Jump-gate
 *
 *
------------------------------------------------------------------*/

enum jumpgate_enum
{
   MAX_JUMPGATES   = 4,
   JUMPGATE_TIME   = 5000,
   JUMPGATE_ANIM   = 250,
   JUMPGATE_FRAMES = 3
};

struct JUMPGATESTRUCT
{
   int  active;
   long time;
   long anim;
   long frame;
   int  dir;
   VECTOR4  pos;
} jgates[MAX_JUMPGATES];

static int jgcur;   /* JUMPGATE index */
static int jcount;  /* JUMPGATE counter */

static VECTOR4 jgvert[32] =
{
   {-10.0f, 0.0f, 10.0f, 1.0f},
   {-0.0f, 10.0f, -10.0f, 1.0f},
   {10.0f, 0.0f, -10.0f, 1.0f},
   {0.0f, -10.0f, 10.0f, 1.0f},

   {-30.0f, 0.0f, 30.0f, 1.0f},
   {0.0f, 30.0f, -30.0f, 1.0f},
   {30.0f, 0.0f, -30.0f, 1.0f},
   {0.0f, -30.0f, 30.0f, 1.0f},

   {-50.0f, 0.0f, 50.0f, 1.0f},
   {0.0f, 50.0f, -50.0f, 1.0f},
   {50.0f, 0.0f, -50.0f, 1.0f},
   {0.0f, -50.0f, 50.0f, 1.0f},

   {-70.0f, 0.0f, 70.0f, 1.0f},
   {0.0f, 70.0f, -70.0f, 1.0f},
   {70.0f, 0.0f, -70.0f, 1.0f},
   {0.0f, -70.0f, 70.0f, 1.0f},

   {-10.0f, 0.0f, -10.0f, 1.0f},
   {0.0f, 10.0f, 10.0f, 1.0f},
   {10.0f, 0.0f, 10.0f, 1.0f},
   {0.0f, -10.0f, -10.0f, 1.0f},

   {-30.0f, 0.0f, -30.0f, 1.0f},
   {0.0f, 30.0f, 30.0f, 1.0f},
   {30.0f, 0.0f, 30.0f, 1.0f},
   {0.0f, -30.0f, -30.0f, 1.0f},

   {-50.0f, 0.0f, -50.0f, 1.0f},
   {0.0f, 50.0f, 50.0f, 1.0f},
   {50.0f, 0.0f, 50.0f, 1.0f},
   {0.0f, -50.0f, -50.0f, 1.0f},

   {-70.0f, 0.0f, -70.0f, 1.0f},
   {0.0f, 70.0f, 70.0f, 1.0f},
   {70.0f, 0.0f, 70.0f, 1.0f},
   {0.0f, -70.0f, -70.0f, 1.0f}
};

void Jumpgate_init ( void )
{
   int i;
   for ( i=0; i<MAX_JUMPGATES; i++ )
   {
      jgates[i].active = FALSE;
      jgates[i].time   = 0;
      jgates[i].anim   = 0;
      jgates[i].frame  = 0;
      jgates[i].dir    = 0;
      Vector_init ( jgates[i].pos );
   }
   jgcur   = 0;
   jcount  = 0;
}

void Jumpgate_open ( VECTOR4 pos, int dir )
{
   if ( jcount > MAX_JUMPGATES-1 ) return;

   jgates[jgcur].active = TRUE;
   jgates[jgcur].time   = 0;
   jgates[jgcur].anim   = 0;
   jgates[jgcur].frame  = 0;
   jgates[jgcur].dir    = dir * 16;
   Vector_copy  ( pos, jgates[jgcur].pos );

   jgcur++;
   if ( jgcur > MAX_JUMPGATES-1 )
      jgcur = 0;

   jcount++;
}

void Jumpgate_animate ( MATRIX4 r )
{
   MATRIX4 tmp_mat;
   VECTOR4 tmp[16];
   int    i, j, p[32], f0;

   for ( i=0; i<MAX_JUMPGATES; i++ )
   {

      if ( jgates[i].active )
      {
         jgates[i].time += gv->msec;
         jgates[i].anim += gv->msec;

         if ( jgates[i].anim > JUMPGATE_ANIM )
         {
            jgates[i].anim -= JUMPGATE_ANIM;
            jgates[i].frame += 1;
            if ( jgates[i].frame > JUMPGATE_FRAMES )
               jgates[i].frame = 0;
         }

         if ( jgates[i].time > JUMPGATE_TIME )
         {
            jgates[i].active = FALSE;
            jcount--;
         }

         /* draw jumpgate */
         Matrix_vec_mult ( r, jgates[i].pos, tmp[0] );
         Matrix_copy ( r, tmp_mat );
         Matrix_set_trans ( tmp_mat, tmp[0] );

         f0 = ( jgates[i].frame + 1 ) * 4;
         Matrix_vec_multn ( tmp_mat,
               &jgvert[jgates[i].dir], tmp, f0 );
         Camera_project_points ( tmp, p, f0 );
         for ( j=0; j<((f0*2)-4); j+=8 )
         {
            Draw_line ( p[0+j], p[1+j], p[2+j], p[3+j], GREEN );
            Draw_line ( p[2+j], p[3+j], p[4+j], p[5+j], GREEN );
            Draw_line ( p[4+j], p[5+j], p[6+j], p[7+j], GREEN );
            Draw_line ( p[6+j], p[7+j], p[0+j], p[1+j], GREEN );
         }
      }
   }
}

/*------------------------------------------------------------------
 * One-up!!!
 *
 *
------------------------------------------------------------------*/

enum oneup_enum
{
   ONEUP_LIFE       = 2000,
   ONEUP_BLEND_TIME = 200
};

struct ONEUPSTRUCT
{
   VECTOR4  pos;
   long frame;
   long blend;
   long color;
   int  active;
}one_up;

static VECTOR4 one_up_vert[10] =
{
   {-40.0f, 20.0f, 0.0f, 1.0f} ,  /* 1 */
   {-40.0f, -20.0f, 0.0f, 1.0f },
   {-20.0f, 20.0f, 0.0f, 1.0f  }, /* U */
   {-20.0f, -20.0f, 0.0f, 1.0f },
   { 10.0f, -20.0f, 0.0f, 1.0f },
   { 10.0f, 20.0f, 0.0f, 1.0f  },
   { 20.0f, 20.0f, 0.0f, 1.0f  }, /* P */
   { 20.0f, -20.0f, 0.0f, 1.0f },
   { 60.0f, 0.0f, 0.0f, 1.0f },
   { 20.0f, 0.0f, 0.0, 1.0f }
};

static VECTOR4 oneup_thrust = { 0.0f, 0.0f, 10.0f, 1.0f };

void One_up_init ( void )
{
   one_up.active = FALSE;
}

void One_up_add  ( OBJECT *obj )
{
   if ( one_up.active == FALSE )
   {
      one_up.active = TRUE;
      Vector_copy ( obj->pos, one_up.pos );
      one_up.frame = 0;
      one_up.blend = 0;
      one_up.color = WHITE;
   }
}

void One_up_draw ( MATRIX4 r )
{
   int p0[20];
   VECTOR4 tmp[10];
   MATRIX4 tmp_mat;

   if ( one_up.active )
   {
      one_up.pos[ZPOS] += oneup_thrust[ZPOS] * gv->fadjust;
      if ( one_up.pos[ZPOS] > -50.0f )
         one_up.active = FALSE;
      one_up.frame += gv->msec;
      if ( one_up.frame > ONEUP_LIFE )
      {
         one_up.active = FALSE;
      }
      one_up.blend += gv->msec;
      if ( one_up.blend > ONEUP_BLEND_TIME )
      {
         one_up.blend -= ONEUP_BLEND_TIME;
         one_up.color -= 3;
      }

      /* draw one_up */
      Matrix_vec_mult ( r, one_up.pos, tmp[0] );
      Matrix_copy ( r, tmp_mat );
      Matrix_set_trans ( tmp_mat, tmp[0] );

      Matrix_vec_multn ( tmp_mat, one_up_vert, tmp, 10 );
      Camera_project_points ( tmp, p0, 10 );

      /* draw 1 */
      Draw_line ( p0[0], p0[1], p0[2], p0[3], one_up.color );

      /* draw U */
      Draw_line ( p0[4], p0[5], p0[6], p0[7], one_up.color );
      Draw_line ( p0[6], p0[7], p0[8], p0[9], one_up.color );
      Draw_line ( p0[8], p0[9], p0[10], p0[11], one_up.color );

      /* draw P */
      Draw_line ( p0[12], p0[13], p0[14], p0[15], one_up.color );
      Draw_line ( p0[12], p0[13], p0[16], p0[17], one_up.color );
      Draw_line ( p0[16], p0[17], p0[18], p0[19], one_up.color );
   }
}


gtext.h
#ifndef XINVADERS3D_GAMETEXT
#define XINVADERS3D_GAMETEXT

/* XInvaders3d game Text */


/* use a null terminator, so code doesn't have to
	be changed in many places. Don't use \n; assume
	a line break suitable to the particular platform
	should be generated after each line. TBB */
char *game_about_info[] =
{
   "Thanks for playing XInvaders 3D.",
   "XInvaders 3D v1.4.0, Copyright 2000 Don Llopis",
   "XInvaders 3D comes with ABSOLUTELY NO WARRANTY.",
   "This is free software, and you are welcome to",
   "redistribute it under certain conditions. See",
   "file named COPYING for more information.",
   "Latest version of the game can be found at:",
   "https://github.com/JoesCat/xinv3d/releases",
   NULL
};

char *game_rules_info[] =
{
   "Keys:",
   "SPACE: Fire Missile",
   "Left Arrow: move left",
   "Right Arrow: move right",
   "Up Arrow: move up",
   "Down Arrow: move down",
   "f: Show frame-rate",
   "p: Pause game",
   "q: reset game",
   "ESC: Quit program",
   NULL
};


/* 
 * Not so nice vector-fonts:
 * perhaps one day, when the universe is kinder to me,
 * I'll implement a nice vector-font system. But
 * until the universe lets me go on my trip,
 * the stuff below stays!
 */
static int A[17] = 
{ 
   4,
   0, 0, 24, 0,
   0, 12, 24, 12,
   0, 0, 0, 24,
   24, 0, 24, 24
};

static int C[13] =
{
   3,
   0, 0, 24, 0,
   0, 0, 0, 24,
   0, 24, 24, 24
};

static int D[13] = 
{
   3,
   0, 0, 0, 24,
   0, 0, 24, 12,
   0, 24, 24, 12
};

static int E[17] =
{
   4,
   0, 0, 24, 0,
   0, 12, 24, 12,
   0, 24, 24, 24,
   0, 0, 0, 24

};

static int EXCLAMATION[33] =
{
   8,
   0, 0, 24, 0,
   5, 19, 15, 19,
   0, 0, 5, 19,
   24, 0, 15, 19,
   5, 25, 15, 25,
   5, 30, 15, 30,
   5, 25, 5, 30,
   15, 25, 15, 30
};

static int F[13] =
{
   3,
   0, 0, 0, 24,
   0, 0, 24, 0,
   0, 12, 24, 12
};

static int G[17] =
{
   4, 
   24, 0, 0, 12,
   0, 12, 24, 24,
   24, 24, 24, 12,
   24, 12, 12, 12
};

static int H[13] =
{
   3,
   0, 0, 0, 24,
   0, 12, 24, 12,
   24, 0, 24, 24
};

static int I[13] =
{
   3,
   0, 0, 24, 0,
   0, 24, 24, 24,
   12, 0, 12, 24
};

static int M[17] =
{
   4,
   0, 0, 0, 24,
   0, 0, 12, 12,
   12, 12, 24, 0,
   24, 0, 24, 24
};

static int N[13] =
{
   3,
   0, 0, 0, 24,
   0, 0, 24, 24,
   24, 0, 24, 24
};

static int O[17] =
{
   4,
   0, 0, 24, 0,
   0, 0, 0, 24,
   0, 24, 24, 24,
   24, 0, 24, 24
};


static int P[13] =
{
   3,
   0, 0, 0, 24,
   0, 0, 24, 12,
   24, 12, 0, 12
};

static int R[17] =
{
   4,
   0, 0, 0, 24,
   0, 0, 24, 12,
   0, 12, 24, 12,
   0, 12, 24, 24
};

static int S[13] =
{
   3,
   24, 0, 0, 5,
   0, 5, 24, 15,
   24, 15, 0, 24
};

static int T[9] =
{
   2,
   0, 0, 24, 0,
   12, 0, 12, 24
};

static int U[13] =
{
   3,
   0, 0, 0, 24,
   0, 24, 24, 24,
   24, 24, 24, 0
};

static int V[9] =
{
   2,
   0, 0, 12, 24,
   24, 0, 12, 24
};

static int X[9] =
{
   2,
   0, 0, 24, 24,
   24, 0, 0,24 
};

static int Y[13] =
{
   3,
   0, 0, 12, 12,
   12, 12, 24, 0,
   12, 12, 12, 24
};

static int ZERO[21] =
{
   5,
   0, 0, 24, 24,
   0, 0, 24, 0,
   0, 24, 24, 24,
   0, 0, 0, 24,
   24, 0, 24, 24
};

static int ONE[5] =
{
   1,
   12, 0, 12, 24
};

static int TWO[21] =
{
   5,
   0, 0, 24, 0,
   24, 0, 24, 12,
   24, 12, 0, 12,
   0, 12, 0, 24,
   0, 24, 24, 24
};

static int THREE[17] =
{
   4,
   0, 0, 24, 0, 
   24, 12, 0, 12,
   0, 24, 24, 24,
   24, 0, 24, 24
};
static int FOUR[13] =
{
   3,
   0, 0, 0, 12,
   0, 12, 24, 12,
   24, 0, 24, 24
};

static int FIVE[21] =
{
   5,
   0, 0, 24, 0,
   0, 0, 0, 12,
   0, 12, 24, 12,
   24, 12, 24, 24,
   24, 24, 0, 24
};

static int SIX[17] =
{
   4,
   0, 0, 0, 24,
   0, 24, 24, 24,
   24, 24, 24, 12,
   24, 12, 0, 12 
};

static int SEVEN[9] = 
{
   2, 
   0, 0, 24, 0, 
   24, 0, 0, 24
};
static int EIGHT[21] =
{
   5,
   0, 0, 24, 0,
   0, 12, 24, 12,
   0, 24, 24, 24,
   0, 0, 0, 24,
   24, 0, 24, 24
};

static int NINE[17] =
{
   4,
   0, 0, 24, 0,
   0, 12, 24, 12,
   0, 0, 0, 12,
   24, 0, 24, 24
};

static int *NUMBER[10] =
{
   &ZERO[0],
   &ONE[0],
   &TWO[0],
   &THREE[0],
   &FOUR[0],
   &FIVE[0],
   &SIX[0],
   &SEVEN[0],
   &EIGHT[0],
   &NINE[0]
};

static int *XINVADERS3D_LOGO[] = 
{ 
   &X[0], &I[0], &N[0], &V[0], &A[0], &D[0], &E[0], &R[0], &S[0], 
   &THREE[0], &D[0], NULL
};

static int *GET[] =
{
   &G[0], &E[0], &T[0], NULL
};

static int *READY[] =
{
   &R[0], &E[0], &A[0], &D[0], &Y[0], &EXCLAMATION[0],
   &EXCLAMATION[0], NULL
};

static int *GAME[] =
{
   &G[0], &A[0], &M[0], &E[0], NULL
};
static int *OVER[] =
{
   &O[0], &V[0], &E[0], &R[0], NULL
};

static int *PAUSED[] =
{
   &P[0], &A[0], &U[0], &S[0], &E[0], &D[0], NULL
};

static int *SCORE[] =
{
   &S[0], &C[0], &O[0], &R[0], &E[0], NULL
};

static int *HI_SCORE[] =
{
   &H[0], &I[0], &S[0], &C[0], &O[0], &R[0], &E[0], NULL
};

static int *PRESS[] =
{
   &P[0], &R[0], &E[0], &S[0], &S[0], NULL
};

static int *FIRE[] =
{
   &F[0], &I[0], &R[0], &E[0], NULL
};

static int *TO[] =
{
   &T[0], &O[0], NULL
};

static int *START[] =
{
   &S[0], &T[0], &A[0], &R[0], &T[0], NULL
};

#endif
